### ISSUE TITLE
[CHA-RL4b7][Room monitoring] Delayed RoomStatus transition to `ATTACHING`

### ISSUE THREAD

**Issue by sacOO7** [ably/ably-chat-js#436]
- As per CHA-RL4b7, `If a room lifecycle operation is not in progress and the channel state is ATTACHING and no transient disconnect timeout exists for the contributor, then a transient disconnect timeout with a 5 second limit is created for the contributor. Upon timeout, the room status is transitioned to ATTACHING, using the reason from the initial channel state change as the error for the transition`

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/CHA-772) by [Unito](https://www.unito.io)

**Comment by sacOO7** [comment]
**Associated Issues:**

- We currently need to manage an explicit attach timer that, when it expires after 5 seconds, sets the `RoomStatus` to `Attaching`.
- This introduces additional complexity because the timer must be cleared when the channel transitions to other states during room monitoring. It also needs to be reset when external operations like `ATTACH`, `DETACH`, or `RELEASE` are triggered.
- The current implementation of `chat-js` assumes that the `ATTACHING` state of a channel is emitted by the `DETACHED` event received from the server.  

https://github.com/ably/ably-chat-js/blob/6d1b04a3a9789aefaa8990fffe4f7a022aae38e3/src/core/room-lifecycle-manager.ts#L277-L293

- However, this isn't always the case. When the connection goes `DISCONNECTED`, the channel transitions from `ATTACHED` to `ATTACHING` if it reconnects within 2 minutes (before entering the `SUSPENDED` state). This behavior is in line with [spec RTL3d](https://sdk.ably.com/builds/ably/specification/main/features/#RTL3d)
- According to RTL3d, channels in the `ATTACHING`, `ATTACHED`, or `SUSPENDED` states must be reattached during reconnection.
- As a result, we are not fully transparent about the `RoomStatus.ATTACHING` state in this common scenario.
- This leads to a bug where the `Room` transitions from `ATTACHED` to `ATTACHED` again (as per CHA-RL4b8), without exposing the intermediate `ATTACHING` state.

**Proposed Solution:**

- Instead of maintaining a separate timer, immediately set `RoomStatus` to `ATTACHING` (if it isn't already in that state).
- If the `ATTACHING` state is triggered by a server-sent `DETACH` event (a rare case), we will still set the `RoomStatus` to `ATTACHING`. If the internal reattach attempt fails, the channel will enter the `SUSPENDED` state as per [spec RTL13b](https://sdk.ably.com/builds/ably/specification/main/features/#RTL13b). As part of the Room Monitoring spec CHA-RL4b9, this will cause the `RoomStatus` to transition to `SUSPENDED` and enter an internal retry loop.
- This change eliminates the need to manage explicit attach timers and clear them during other operations.
- It also makes the `RoomStatus.ATTACHING` state more transparent and fixes the bug where the `Room` incorrectly transitions from `ATTACHED` to `ATTACHED` during reconnection within 2 minutes.
- As for handling feature discontinuities, users should rely on the callbacks registered for the respective room features, rather than tracking changes in `RoomStatus`.

**Comment by AndyTWF** [comment]
If I'm understanding you correctly - the inclusion of this timer was by design.

If the library temporarily goes from `ATTACHED` to `ATTACHING` to `ATTACHED` again in a short period with no errors of continuity, does it actually benefit the user to tell them that this has happened? The purpose of the timer was to make it so that transient state changes are handled internally without the user needing to take action.

Now, if we emit `ATTACHED` twice then I agree that's not so ideal - but I think not emitting the intermediary state is as intended.

> When the connection goes DISCONNECTED, the channel transitions from ATTACHED to ATTACHING

Not quite - RTL3d is when connection goes to `CONNECTED`, RTL3e says that no changes to channel state are associated with the `DISCONNECTED state.

**Comment by sacOO7** [comment]
- I agree there are no channel state changes as per [RTL3e](https://sdk.ably.com/builds/ably/specification/main/features/#RTL3e)
- But as per `RTL3d` channel transitions from `ATTACHED` to `ATTACHING`, but `RTL3e` is prerequisite to make it happen

**Comment by sacOO7** [comment]
> Now, if we emit ATTACHED twice then I agree that's not so ideal - but I think not emitting the intermediary state is as intended.

I strongly disagree : (

- `RoomStatus` reflects the aggregated state of the underlying channels. If one or more channels transition to the `ATTACHING` state, this should immediately be reflected as a change in the `RoomStatus` (assuming no other operations are in progress).
- I don't see any negative consequences to being transparent about the status change. Additionally, transitioning the room status from `ATTACHED` to `ATTACHED` could be seen as a bug from the user's perspective.
- Introducing an attach timer complicates the implementation and could jeopardize the atomicity of ongoing room operations (if not managed correctly). Moreover, active timers could create challenges for future implementations as well.

**Comment by sacOO7** [comment]
- It's important to note that while `RoomStatus.Attaching` doesn't directly indicate message loss, the user can use it to display room status on the UI.
- This implies that when we emit `RoomStatus.ATTACHING` event, messages will not be immediately sent/received across any features,so it won't be seen as a bug. 
- I'm considering the scenario where as per current impl.m the `Room` is in the `ATTACHED` state, but no messages are being sent or received ( because channels are internally being `ATTACHED` )

**Comment by sacOO7** [comment]
- This also seems to introduce a bug where features will not adhere to spec [CHA-RL9](https://sdk.ably.com/builds/ably/specification/main/chat-features/#:~:text=(-,CHA%2DRL9,-)%20Many%20operations)
- Since we are delaying the transition to `RoomStatus.ATTACHING`, features will not wait for `Room` to get `ATTACHED` to perform the given operation, instead, they will treat the room as`ATTACHED` and messages will either be queued or rejected internally : (
- As per CHA-RL9a, caller should wait till `RoomStatus` is `ATTACHED`,  for other RoomStatus change it returns error with `RoomInInvalidState` code.
- Currently, operation will not return with `error` even when `Room` goes into `Suspended` state after `ATTACHING` fails : (
- So, current implementation doesn't conform to spec [CHA-RL9](https://sdk.ably.com/builds/ably/specification/main/chat-features/#:~:text=(-,CHA%2DRL9,-)%20Many%20operations) : (

**Comment by AndyTWF** [comment]
> Introducing an attach timer complicates the implementation and could jeopardize the atomicity of ongoing room operations (if not managed correctly)

I don't think we should be potentially passing complexity onto users just because it requires more careful handling internally - surely the idea is that we take on the complexity and deal with it once so that users don't have to?

At the time of conception, CHA-RL9 wasn't a thing (which appreciate changes things), and this was the least-work-for-users approach at the time.

> I'm considering the scenario where as per current impl. the Room is in the ATTACHED state, but no messages are being sent or received immediately ( because channels internally being ATTACHED), unfortunately, this will be seen as a bug

That's why the timeout is "transient" - its short enough that not receiving a couple of messages in that period won't be seen as a bug.

> Delayed RoomStatus transition to ATTACHING also seems to introduce a bug where room features will not adhere to the spec

I agree that CHA-RL9 changes things and that we should look into it. My personal thought is that this shouldn't block the beta releases as this is an edge case, but is something we should address going forwards.

> So, current implementation doesn't conform to spec

Yes - we added CHA-RL9 later on and have a ticket to conform to it at a later date (it's just not the highest priority at the moment).

**Comment by AndyTWF** [comment]
This has now been resolved due to a redesign of how room lifecycle works.
